test/extensions/filters/http/common/aws/signer_impl_test.cc

```diff
@@ -0,0 +1,193 @@
+  EXPECT_CALL(*credentials_provider_, getCredentials()).WillOnce(Return(Credentials()));
+  EXPECT_CALL(*region_provider_, getRegion()).Times(0);
+  signer_.sign(*message_);
+  EXPECT_EQ(nullptr, message_->headers().Authorization());
+}
+
+TEST_F(SignerImplTest, MissingRegionException) {
+  absl::optional<std::string> no_region;
+  EXPECT_CALL(*credentials_provider_, getCredentials()).WillOnce(Return(credentials_));
+  EXPECT_CALL(*region_provider_, getRegion()).WillOnce(Return(no_region));
+  EXPECT_THROW_WITH_MESSAGE(signer_.sign(*message_), EnvoyException,
+                            "Could not determine AWS region");
+  EXPECT_EQ(nullptr, message_->headers().Authorization());
+}
+
+TEST_F(SignerImplTest, SecurityTokenHeader) {
+  EXPECT_CALL(*credentials_provider_, getCredentials()).WillOnce(Return(token_credentials_));
+  EXPECT_CALL(*region_provider_, getRegion()).WillOnce(Return(region_));
+  addMethod("GET");
+  addPath("/");
+  signer_.sign(*message_);
+  EXPECT_STREQ("token", message_->headers().get(SignerImpl::X_AMZ_SECURITY_TOKEN)->value().c_str());
+  EXPECT_STREQ("AWS4-HMAC-SHA256 Credential=akid/20180102/region/service/aws4_request, "
+               "SignedHeaders=x-amz-content-sha256;x-amz-date;x-amz-security-token, "
+               "Signature=1d42526aabf7d8b6d7d33d9db43b03537300cc7e6bb2817e349749e0a08f5b5e",
+               message_->headers().Authorization()->value().c_str());
+}
+
+TEST_F(SignerImplTest, DateHeader) {
+  EXPECT_CALL(*credentials_provider_, getCredentials()).WillOnce(Return(credentials_));
+  EXPECT_CALL(*region_provider_, getRegion()).WillOnce(Return(region_));
+  addMethod("GET");
+  addPath("/");
+  signer_.sign(*message_);
+  EXPECT_STREQ("20180102T030400Z",
+               message_->headers().get(SignerImpl::X_AMZ_DATE)->value().c_str());
+  EXPECT_STREQ("AWS4-HMAC-SHA256 Credential=akid/20180102/region/service/aws4_request, "
+               "SignedHeaders=x-amz-content-sha256;x-amz-date, "
+               "Signature=4ee6aa9355259c18133f150b139ea9aeb7969c9408ad361b2151f50a516afe42",
+               message_->headers().Authorization()->value().c_str());
+}
+
+TEST_F(SignerImplTest, EmptyContentHeader) {
+  EXPECT_CALL(*credentials_provider_, getCredentials()).WillOnce(Return(credentials_));
+  EXPECT_CALL(*region_provider_, getRegion()).WillOnce(Return(region_));
+  addMethod("GET");
+  addPath("/empty?content=none");
+  signer_.sign(*message_);
+  EXPECT_STREQ("e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
+               message_->headers().get(SignerImpl::X_AMZ_CONTENT_SHA256)->value().c_str());
+  EXPECT_STREQ("AWS4-HMAC-SHA256 Credential=akid/20180102/region/service/aws4_request, "
+               "SignedHeaders=x-amz-content-sha256;x-amz-date, "
+               "Signature=999e211bc7134cc685f830a332cf4d871b6d8bb8ced9367c1a0b59b95a03ee7d",
+               message_->headers().Authorization()->value().c_str());
+}
+
+TEST_F(SignerImplTest, ContentHeader) {
```

Comment:
Can you add short `//` explanations of what each of these tests does above each `TEST_F`?

Reply:
Will do.